processing/src/main/java/org/apache/druid/error/ErrorResponse.java

```diff
@@ -0,0 +1,214 @@
+      // and helps aid the removal of QueryException.
+      failure = new QueryExceptionCompat(
+          new QueryException(
+              nullOrString(map.get("error")),
+              nullOrString(map.get("errorMessage")),
+              nullOrString(map.get("errorClass")),
+              nullOrString(map.get("host"))
+          )
+      );
+    } else {
+      failure = new DruidException.Failure(stringOrFailure(map, "errorCode"))
+      {
+        @Override
+        protected DruidException makeException(DruidException.DruidExceptionBuilder bob)
+        {
+          final DruidException retVal = bob.forPersona(DruidException.Persona.valueOf(stringOrFailure(map, "persona")))
+                                           .ofCategory(DruidException.Category.valueOf(stringOrFailure(
+                                               map,
+                                               "category"
+                                           )))
+                                           .build(stringOrFailure(map, "errorMessage"));
+
+          final Object context = map.get("context");
+          if (context instanceof Map) {
+            //noinspection unchecked
+            retVal.withContext((Map<String, String>) context);
+          }
+
+          return retVal;
+        }
+      };
+    }
+    return new ErrorResponse(DruidException.fromFailure(new DeserializedFailure(failure)));
+  }
+
+  private final DruidException underlyingException;
+
+  public ErrorResponse(DruidException underlyingException)
+  {
+    this.underlyingException = underlyingException;
+  }
+
+  @JsonValue
+  public Map<String, Object> getAsMap()
+  {
+    final LinkedHashMap<String, Object> retVal = new LinkedHashMap<>();
+
+    // This if statement is a compatibility layer to help bridge the time while we are introducing the DruidException.
+    // In a future release, QueryException should be completely eliminated, at which point we should also be
+    // able to eliminate this compatibility layer.
+    if (QueryExceptionCompat.ERROR_CODE.equals(underlyingException.getErrorCode())) {
+      retVal.put("error", underlyingException.getContextValue("legacyErrorCode"));
+      retVal.put("errorClass", underlyingException.getContextValue("errorClass"));
+      retVal.put("host", underlyingException.getContextValue("host"));
+    } else {
+      retVal.put("error", "druidException");
+    }
+
+    retVal.put("errorCode", underlyingException.getErrorCode());
+    retVal.put("persona", underlyingException.getTargetPersona().toString());
+    retVal.put("category", underlyingException.getCategory().toString());
+    retVal.put("errorMessage", underlyingException.getMessage());
+    retVal.put("context", underlyingException.getContext());
+
+    return retVal;
+  }
+
+  public DruidException getUnderlyingException()
+  {
+    return underlyingException;
+  }
+
+  @Nullable
+  private static String nullOrString(Object o)
```

Comment:
there is a similar existing method, `Evals.asString` though its primarily used for processing data for expressions and query time stuff, so i don't feel super strongly about using it here or not